### DISCUSSION TITLE
Schema auto-inferred

### DISCUSSION THREAD

**Discussion by calebAtIspot** [terraform-linters/tflint-ruleset-opa#94] [Ideas]
It would be nice if the schema was auto-inferred from the resource so you don't have to manually specify it. In a Ideal world if I had a resource like 

```hcl
data "aws_iam_policy_document" "example" {
  statement {
    sid    = "bad-sid-dashes-not-allowed"
    effect = "Allow"
    actions = [
      "foo"
    ]
    resources = [
      "fah"
    ]
  }
}

```

then I could just do `terraform.data_sources("aws_ami")` instead of `terraform.data_sources("aws_ami", {"statement": {"sid": "string"}}, {})`

**Comment by wata727** [comment]
Thank you for raising this idea.

I agree that explicit schema declaration feels redundant, but HCL is a syntax-agnostic, schema-driven language, and there is no standard mechanism for automatically inferring schemas.
https://github.com/hashicorp/hcl/blob/v2.20.0/spec.md

To achieve this, the following approaches may be considered:

1. Fork HCL to implement your own schema inference
2. Include the schemas of all resources in this ruleset or add dependencies on the Terraform providers so they can be retrieved dynamically
3. Parse the Rego query language to calculate referenced schemas

My current understanding is that these become too complex, so we're taking the current approach of forcing explicit schema declarations.
